### PR TITLE
Some changes from STT work

### PR DESCRIPTION
**Pull request by reuben**

### CODE REVIEW

File: coqpit/coqpit.py [hunk 1]
```diff
@@ -278,6 +280,27 @@ def _deserialize(x: Any, field_type: Any) -> Any:
...
+def rgetattr(obj, attr, *args):
+    def _getitem(obj, attr):
+        return operator.getitem(obj, int(attr), *args)
+
+    def _getattr(obj, attr):
+        getfunc = _getitem if attr.isnumeric() else getattr
+        return getfunc(obj, attr, *args)
+    return functools.reduce(_getattr, [obj] + attr.split('.'))
+
+
```
Comment (erogol) [inline coqpit/coqpit.py:303]: I don't think this part correctly handles fields inherited from the parent class. 

You can try this. 
https://gist.github.com/erogol/23dc5cc3deb8fae714a98bbdcbdc370d

Comment (reuben) [reply coqpit/coqpit.py:303]: That script passes for me with the changes in this PR.

File: coqpit/coqpit.py [hunk 2]
```diff
@@ -608,56 +643,46 @@ def load_json(self, file_name: str) -> None:
...
             Coqpit: new config object with updated values.
         """
+        if not args:
+            parser = self.init_argparse()
+            args = parser.parse_args()
         if isinstance(args, argparse.Namespace):
             args_dict = vars(args)
         elif isinstance(args, list):
             # overriding values from .parse_known_args()
             args_dict = dict([*zip(args[::2], args[1::2])])
```
Comment (erogol) [inline coqpit/coqpit.py:662]: One issue is there when you give a list as input. For instance, the output of `parser.parse_known_args()` is a list and `--` needs to be replaced as in the old code.

Comment (reuben) [reply coqpit/coqpit.py:662]: I don't see how that's connected to this PR?

Comment (reuben) [reply coqpit/coqpit.py:662]: Oh, nevermind. It's just a different change. I get it.

**Comment by erogol** [comment]
Another issue is that the prefix you used for False `bool` values come before `coqpit`. 

```python
@dataclass
class Bla():
    val_bool: bool
```
converts to `--coqpit. val_bool --nocoqpit. val_bool`

IMO it is better to keep a single argument and take `true`  or `false` as input

File: coqpit/coqpit.py [hunk 3]
```diff
@@ -608,56 +643,46 @@ def load_json(self, file_name: str) -> None:
...
         Args:
-            args (namespace): argeparse namespace as an output of ```argparse.parse_arguments()```.
+            args (namespace, optional): argeparse namespace as an output of ```argparse.parse_arguments()```. If unspecified will use a newly created parser with ```init_argparse()```.
 
         Returns:
             Coqpit: new config object with updated values.
         """
+        if not args:
+            parser = self.init_argparse()
+            args = parser.parse_args()
```
Comment (erogol) [inline coqpit/coqpit.py:658]: This breaks when there is no argument for Coqpit but arguments for other parts and those arguments raise `undefined argument`

**Comment by erogol** [comment]
I feel like keeping the old lazy `parse_argparse` and letting the user match the expected value types (either with `check_values()` or just checking the expected type by looking to the cockpit class) is a better option. 

Using argparse creates a lot of edge cases that are hard to fill.  I don't know if we have time and enough motivation for it. 

But I guess the problem with STT is that the users are reliant on console arguments more than TTS users so type checking makes a difference.  ??

**Comment by reuben** [comment]
Sorry I don't understand your comment at all. What edge cases?

**Comment by reuben** [comment]
I've changed the handling for boolean flags, so now all comments should be addressed! Let me know if you have a test case for the `Union` thing.
